### ISSUE TITLE
🌀[API] Ye Olde TypeScript Refactor

### ISSUE THREAD

**Issue by emplums** [primer/components#545]
After much thought, I think it would be a good move to refactor our components using TypeScript. 

My main hesitation was that components written in TS might be a barrier to contributors who aren't as comfortable with TS, but after some thought, it seems like the main users of Primer Components are at GitHub already using TS in their project or have some TS knowledge. 

Right now, we've got typings in a separate file in order for TS projects to be able to use Components easily. This leads to a maintenance nightmare because components frequently get updated without their typings getting updated, and because we aren't actually compiling TS anywhere, we aren't catching it.

We should be able to start doing this incrementally - as we use the components in the UI exploration project we can start updating them here ✨

Let's use this issue as a tracking issue, and check off each component refactored + link to the PR as we go. 🎉

### Tracking

- [ ] Avatar
- [ ] BorderBox
- [ ] Box
- [ ] BranchName
- [ ] Button
- [ ] CircleBadge
- [ ] CircleOcticon
- [ ] CounterLabel
- [ ] Details
- [ ] Dialog
- [ ] Dropdown
- [ ] FilterList
- [ ] Flash
- [ ] Flex
- [ ] Grid
- [ ] Heading
- [ ] Label
- [ ] Link
- [ ] PointerBox
- [ ] Position
- [ ] StateLabel
- [ ] StyledOcticon
- [ ] TabNav
- [ ] TextInput
- [ ] Text
- [ ] Tooltip
- [ ] UnderlineNav

**Comment by T-Hugs** [comment]
Good way to get started is by adding a tsconfig.json with `"allowJs: true"`, then add it to the build. Then components can be updated one by one.

**Comment by dmarcey** [comment]
@emplums Would there be any issue with moving to nextjs 9.x? It looks like we're currently on
`"next": "7.0.2",`.

I was just searching around for TypeScript support with nextjs and stumbled across this blog post: 
https://nextjs.org/blog/next-9#built-in-zero-config-typescript-support

If there isn't any opposition to upgrading us, I'll try to get the configuration all set up and convert the first component.

**Comment by shawnbot** [comment]
@dmarcey Next.js will go away entirely with #522 😁

**Comment by dmarcey** [comment]
In that case, I vote we wait for that PR before we start on the TS conversion! 🚧

**Comment by emplums** [comment]
Should go out tomorrow! 🎉

**Comment by colebemis** [comment]
Closing in favor of https://github.com/primer/components/issues/970
